memcpy buffer overflow: load_construct_data() and std::complex members

I get a weird compilation warning and a subsequent runtime failure (buffer overflow) from the following code:
```c++
#include <complex>
#include <iostream>

#include <hpx/runtime/serialization/complex.hpp>
#include <hpx/runtime/serialization/vector.hpp>

struct my_struct {

    using x_type = std::complex<double>;

    my_struct() = delete;
    my_struct(x_type x) : x(x) {}

    bool operator==(my_struct const& ms) const { return x == ms.x; }

    x_type x;

    /// HPX.Serialization
    template<typename Archive>
    inline void serialize(Archive & ar, const unsigned int) {
        ar & x;
    }
    template <class Archive>
    inline void friend load_construct_data(Archive & ar, my_struct* b, const unsigned int) {
        ::new(b) my_struct(0);
    }
};

int main(int argc, char **argv) {

    std::vector<char> buf;

    // Serialize
    std::vector<my_struct> in;
    in.push_back(my_struct(1.0));
    in.push_back(my_struct(2.0));
    hpx::serialization::output_archive oa(buf);
    oa << in;

    // Deserialize
    std::vector<my_struct> out;
    hpx::serialization::input_archive ia(buf);
    ia >> out;

    if(!(in == out)) {
        std::cout << "Serialization failure" << std::endl;
        return -1;
    }
    return 0;
}
```
The warning reads
```
/usr/include/bits/string3.h:53:71: warning: call to void* __builtin___memcpy_chk(void*, const void*, long unsigned int, long unsigned int) will always overflow destination buffer
   return __builtin___memcpy_chk (__dest, __src, __len, __bos0 (__dest));
```
valgrind reveals the point of the failure to be in `serialization_access_data.hpp:159`
```
**14384** *** memcpy_chk: buffer overflow detected ***: program terminated
==14384==    at 0x4C2EB04: VALGRIND_PRINTF_BACKTRACE (valgrind.h:6818)
==14384==    by 0x4C3363A: __memcpy_chk (vg_replace_strmem.c:1574)
==14384==    by 0x404D33: memcpy (string3.h:53)
==14384==    by 0x404D33: read (serialization_access_data.hpp:159)
==14384==    by 0x404D33: load_binary (input_container.hpp:107)
==14384==    by 0x404D33: load_binary (input_archive.hpp:243)
==14384==    by 0x404D33: load_bitwise<std::complex<double> > (input_archive.hpp:156)
==14384==    by 0x404D33: load<std::complex<double> > (input_archive.hpp:89)
==14384==    by 0x404D33: invoke_impl<std::complex<double> > (input_archive.hpp:78)
==14384==    by 0x404D33: invoke<std::complex<double> > (basic_archive.hpp:66)
==14384==    by 0x404D33: operator&<std::complex<double> > (serialize.hpp:43)
==14384==    by 0x404D33: serialize<hpx::serialization::input_archive> (hpx_bug.cpp:21)
==14384==    by 0x404D33: call<hpx::serialization::input_archive> (access.hpp:88)
==14384==    by 0x404D33: serialize<hpx::serialization::input_archive, my_struct> (access.hpp:113)
==14384==    by 0x404D33: load_nonintrusively_polymorphic<my_struct> (input_archive.hpp:163)
==14384==    by 0x404D33: load_bitwise<my_struct> (input_archive.hpp:141)
==14384==    by 0x404D33: load<my_struct> (input_archive.hpp:89)
==14384==    by 0x404D33: invoke_impl<my_struct> (input_archive.hpp:78)
==14384==    by 0x404D33: invoke<my_struct> (basic_archive.hpp:66)
==14384==    by 0x404D33: operator>><my_struct> (serialize.hpp:29)
==14384==    by 0x404D33: call<hpx::serialization::input_archive, std::vector<my_struct> > (serialize_collection.hpp:100)
==14384==    by 0x404D33: load_collection<hpx::serialization::input_archive, std::vector<my_struct> > (serialize_collection.hpp:124)
==14384==    by 0x404D33: load_impl<my_struct, std::allocator<my_struct> > (vector.hpp:34)
==14384==    by 0x404D33: serialize<my_struct, std::allocator<my_struct> > (vector.hpp:88)
==14384==    by 0x404D33: serialize_force_adl<std::vector<my_struct> > (access.hpp:33)
==14384==    by 0x404D33: call<hpx::serialization::input_archive> (access.hpp:69)
==14384==    by 0x404D33: serialize<hpx::serialization::input_archive, std::vector<my_struct> > (access.hpp:113)
==14384==    by 0x404D33: load_nonintrusively_polymorphic<std::vector<my_struct> > (input_archive.hpp:163)
==14384==    by 0x404D33: load_bitwise<std::vector<my_struct> > (input_archive.hpp:141)
==14384==    by 0x404D33: load<std::vector<my_struct> > (input_archive.hpp:89)
==14384==    by 0x404D33: invoke_impl<std::vector<my_struct> > (input_archive.hpp:78)
==14384==    by 0x404D33: invoke<std::vector<my_struct> > (basic_archive.hpp:66)
==14384==    by 0x404D33: operator>><std::vector<my_struct> > (serialize.hpp:29)
==14384==    by 0x404D33: main (hpx_bug.cpp:43)
```
The most interesting part here is that if I change `x_type` to `double`, the program compiles and runs perfectly.

--- a/hpx/runtime/serialization/detail/serialize_collection.hpp
+++ b/hpx/runtime/serialization/detail/serialize_collection.hpp
@@ -86,8 +86,8 @@ namespace hpx { namespace serialization { namespace detail {
                     typename Collection::size_type size)
             {
                 using value_type = typename Collection::value_type;
-                using storage_type = std::aligned_storage<
-                    sizeof(value_type), alignof(value_type)>;
+                using storage_type = typename std::aligned_storage<
+                    sizeof(value_type), alignof(value_type)>::type;
 
                 collection.clear();
                 hpx::traits::detail::reserve_if_reservable(collection, size);

--- a/tests/regressions/CMakeLists.txt
+++ b/tests/regressions/CMakeLists.txt
@@ -1,5 +1,5 @@
 # Copyright (c) 2011-2012 Bryce Adelstein-Lelbach
-# Copyright (c) 2007-2015 Hartmut Kaiser
+# Copyright (c) 2007-2017 Hartmut Kaiser
 #
 # Distributed under the Boost Software License, Version 1.0. (See accompanying
 # file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
@@ -15,6 +15,7 @@ set(subdirs
     lcos
     parallel
     performance_counters
+    serialization
     threads
     traits
     util

--- a/tests/regressions/serialization/CMakeLists.txt
+++ b/tests/regressions/serialization/CMakeLists.txt
@@ -1,0 +1,37 @@
+# Copyright (c) 2017 Hartmut Kaiser
+#
+# Distributed under the Boost Software License, Version 1.0. (See accompanying
+# file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+set(tests
+    buffer_overrun_2839
+   )
+
+foreach(test ${tests})
+  set(sources
+      ${test}.cpp)
+
+  source_group("Source Files" FILES ${sources})
+
+  # add example executable
+  add_hpx_executable(${test}_test
+                     SOURCES ${sources}
+                     ${${test}_FLAGS}
+                     EXCLUDE_FROM_ALL
+                     HPX_PREFIX ${HPX_BUILD_PREFIX}
+                     FOLDER "Tests/Regressions/Serialization")
+
+  add_hpx_regression_test("serialization" ${test} ${${test}_PARAMETERS})
+
+  # add a custom target for this test
+  add_hpx_pseudo_target(tests.regressions.serialization_dir.${test})
+
+  # make pseudo-targets depend on master pseudo-target
+  add_hpx_pseudo_dependencies(tests.regressions.serialization_dir
+                              tests.regressions.serialization_dir.${test})
+
+  # add dependencies to pseudo-target
+  add_hpx_pseudo_dependencies(tests.regressions.serialization_dir.${test}
+                              ${test}_test_exe)
+endforeach()
+

--- a/tests/regressions/serialization/buffer_overrun_2839.cpp
+++ b/tests/regressions/serialization/buffer_overrun_2839.cpp
@@ -1,0 +1,64 @@
+//  Copyright (c) 2017 Igor Krivenko
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#include <hpx/hpx_main.hpp>
+#include <hpx/runtime/serialization/complex.hpp>
+#include <hpx/runtime/serialization/vector.hpp>
+#include <hpx/util/lightweight_test.hpp>
+
+#include <complex>
+#include <vector>
+
+struct my_struct
+{
+    using x_type = std::complex<double>;
+
+    my_struct() = delete;
+    my_struct(x_type x)
+      : x(x)
+    {
+    }
+
+    bool operator==(my_struct const& ms) const
+    {
+        return x == ms.x;
+    }
+
+    x_type x;
+
+    /// HPX.Serialization
+    template <typename Archive>
+    inline void serialize(Archive& ar, const unsigned int)
+    {
+        ar & x;
+    }
+    template <class Archive>
+    inline void friend load_construct_data(
+        Archive& ar, my_struct* b, const unsigned int)
+    {
+        ::new (b) my_struct(0);
+    }
+};
+
+int main(int argc, char** argv)
+{
+    std::vector<char> buf;
+
+    // Serialize
+    std::vector<my_struct> in;
+    in.push_back(my_struct(std::complex<double>(1.0)));
+    in.push_back(my_struct(std::complex<double>(2.0)));
+    hpx::serialization::output_archive oa(buf);
+    oa << in;
+
+    // Deserialize
+    std::vector<my_struct> out;
+    hpx::serialization::input_archive ia(buf);
+    ia >> out;
+
+    HPX_TEST(in == out);
+
+    return hpx::util::report_errors();
+}

